lib/Basics/process-utils.cpp

```diff
@@ -286,6 +296,127 @@ static void StartExternalProcess(ExternalProcess* external, bool usePipes,
     }
   }
 
+  int err = 0;  // accumulate any errors we might get
+  posix_spawn_file_actions_t file_actions;
+  err |= posix_spawn_file_actions_init(&file_actions);
+  // file actions are performed in order they were added.
+
+  if (usePipes) {
+    err |= posix_spawn_file_actions_adddup2(&file_actions,
+                                            pipe_server_to_child[0], 0);
+    err |= posix_spawn_file_actions_adddup2(&file_actions,
+                                            pipe_child_to_server[1], 1);
+
+    err |= posix_spawn_file_actions_addclose(&file_actions,
+                                             pipe_server_to_child[0]);
+    err |= posix_spawn_file_actions_addclose(&file_actions,
+                                             pipe_server_to_child[1]);
+    err |= posix_spawn_file_actions_addclose(&file_actions,
+                                             pipe_child_to_server[0]);
+    err |= posix_spawn_file_actions_addclose(&file_actions,
+                                             pipe_child_to_server[1]);
+  } else {
+    err |= posix_spawn_file_actions_addopen(&file_actions, 0, "/dev/null",
+                                            O_RDONLY, 0);
+  }
+
+  if (!fileForStdErr.empty()) {
+    err |= posix_spawn_file_actions_addopen(&file_actions, 2,
+                                            fileForStdErr.c_str(),
+                                            O_CREAT | O_WRONLY | O_TRUNC, 0644);
+  }
+
+  posix_spawnattr_t spawn_attrs;
+  err |= posix_spawnattr_init(&spawn_attrs);
+
+  if (err != 0) {
+    external->_status = TRI_EXT_PIPE_FAILED;
+    if (usePipes) {
+      close(pipe_server_to_child[0]);
+      close(pipe_server_to_child[1]);
+      close(pipe_child_to_server[0]);
+      close(pipe_child_to_server[1]);
+    }
+    return;
+  }
+
+  std::vector<char*> envs;
+  for (char** e = environ; *e != nullptr; ++e) {
+    envs.push_back(*e);
+  }
+
+  envs.reserve(envs.size() + additionalEnv.size() + 1);
+  std::transform(
+      additionalEnv.begin(), additionalEnv.end(), std::back_inserter(envs),
+      [](auto& str) -> char* { return const_cast<char*>(str.data()); });
+  envs.emplace_back(nullptr);
+
+  int result = posix_spawnp(&external->_pid, external->_executable.c_str(),
+                            &file_actions, &spawn_attrs, external->_arguments,
+                            envs.data());
+
+  ScopeGuard cleanup([&]() noexcept {
```

Comment:
Doing it here is too late. The scope guard should be moved up to line 331.
Otherwise we won't free the attributes in case we are returning in line 340.

Reply:
Have moved it up.